### PR TITLE
For loop edges, try to avoid overlap with other edges

### PR DESCRIPTION
**Pull request by josejulio**
** Describe the change **

It tries to move the loop edges when it finds overlaps with other edges

Fixes kiali/kiali#1610

** Screenshot **

Before:

Screenshot_2019-10-14_11-39-54

After:
Screenshot_2019-10-14_11-21-38

Before:
Screenshot_2019-10-14_11-39-36

After:

Screenshot_2019-10-14_11-21-56

### CODE REVIEW

**Comment by josejulio** [comment]
Added some comments to explain the flow of the algorithm. Please let me know if that helps.

File: src/components/CytoscapeGraph/CytoscapeGraph.tsx [hunk 1]
```diff
@@ -859,6 +861,98 @@ export class CytoscapeGraph extends React.Component<CytoscapeGraphProps, Cytosca
...
+      otherEdges.forEach(edge => {
+        const testPoint = edge.source().same(node) ? edge.sourceEndpoint() : edge.targetEndpoint();
+        if (
+          squaredDistance(testPoint, loop.sourceEndpoint()) <= minDistance ||
+          squaredDistance(testPoint, loop.targetEndpoint()) <= minDistance
+        ) {
+          found = true;
+          return false; // break the cytoscape forEach
+        }
+        return true;
```
Comment (jshaughn) [inline src/components/CytoscapeGraph/CytoscapeGraph.tsx]: Is this line necessary?

File: src/components/CytoscapeGraph/CytoscapeGraph.tsx [hunk 2]
```diff
@@ -859,6 +861,98 @@ export class CytoscapeGraph extends React.Component<CytoscapeGraphProps, Cytosca
...
+      let found = false;
+      // Check if we have any other edge that overlaps with any of our loop edges
+      otherEdges.forEach(edge => {
+        const testPoint = edge.source().same(node) ? edge.sourceEndpoint() : edge.targetEndpoint();
+        if (
+          squaredDistance(testPoint, loop.sourceEndpoint()) <= minDistance ||
+          squaredDistance(testPoint, loop.targetEndpoint()) <= minDistance
+        ) {
+          found = true;
+          return false; // break the cytoscape forEach
```
[the remark this answers is not in the document]
Comment (josejulio) [reply src/components/CytoscapeGraph/CytoscapeGraph.tsx]: sure

File: src/components/CytoscapeGraph/CytoscapeGraph.tsx (hunk 1 above)
Comment (josejulio) [reply src/components/CytoscapeGraph/CytoscapeGraph.tsx]: I got a "not all code paths return a value" error, but could update the function to signal that it returns false or undefined

**Comment by josejulio** [comment]
> If I understand, it looks for the largest open arc and tries to use it if possible

That's correct!

**Comment by josejulio** [comment]
> Just a couple of comment tweaks and I think it's good.

Done!

File: src/components/CytoscapeGraph/CytoscapeGraph.tsx (hunk 1 above)
Comment (jshaughn) [reply src/components/CytoscapeGraph/CytoscapeGraph.tsx]: ok

File: src/components/CytoscapeGraph/CytoscapeGraph.tsx [hunk 3]
```diff
@@ -859,6 +861,99 @@ export class CytoscapeGraph extends React.Component<CytoscapeGraphProps, Cytosca
...
+      // Default values in rads (taken from cytoscape docs)
+      const DEFAULT_LOOP_SWEEP = -1.5707;
+      const DEFAULT_LOOP_DIRECTION = -0.7854;
+
+      loop.style('loop-direction', DEFAULT_LOOP_DIRECTION);
+      loop.style('loop-sweep', DEFAULT_LOOP_SWEEP);
+
+      let found = false;
+      // Check if we have any other edge that overlaps with any of our loop edges
+      // this uses cytoscape forEach (https://js.cytoscape.org/#eles.forEach)
```
Comment (jshaughn) [inline src/components/CytoscapeGraph/CytoscapeGraph.tsx]: The outer loop also uses eles.forEach but no big deal.

**Pull request merged**